crates/accelerate/src/synthesis/discrete_basis/math.rs

```diff
@@ -0,0 +1,272 @@
+    if f(a).abs() < tol {
+        return Some(a);
+    }
+    if f(b).abs() < tol {
+        return Some(b);
+    }
+    if f(a) * f(b) > ZERO {
+        return None; // invalid boundaries
+    }
+
+    let half = BigFloat::from_f64(0.5);
+    let mut midpoint = half * (a + b);
+    for _ in 0..maxiter {
+        // solution is in [a, midpoint]
+        if f(a) * f(midpoint) < ZERO {
+            b = midpoint;
+        } else {
+            a = midpoint;
+        }
+        midpoint = half * (a + b);
+
+        if (b - a) < tol {
+            break; // we're done!
+        }
+    }
+    Some(midpoint)
+}
+
+pub(crate) fn solve_decomposition_angle(matrix: &Matrix3<BigFloat>) -> BigFloat {
+    let trace = matrix.trace().min(&BigFloat::from_f64(3.0)); // avoid roundoff errors
+    let angle = ((trace - ONE) / TWO).acos();
+    let lhs = (angle / TWO).sin();
+
+    // we use a bisect search to solve the angle equation (Eq. 10 in the paper)
+    let min_angle = ZERO;
+    let max_angle = TWO * (ONE / TWO.powf(BigFloat::from_f64(0.25))).asin();
+
+    let f = move |phi: BigFloat| -> BigFloat {
+        let sin_sq = (phi / TWO).sin().powi(2);
+        TWO * sin_sq * (ONE - sin_sq.powi(2)).sqrt() - lhs
+    };
+
+    bisect(f, min_angle, max_angle, BigFloat::from_f64(1e-20), 70)
+        .expect("Who coded this cannot math")
+}
+
+/// Add coeff * V to the out matrix, where V is the skew-symmetric representation of the
+/// cross product with the vector v.
+fn add_cross_prod(coeff: BigFloat, v: &Matrix3x1<BigFloat>, out: &mut Matrix3<BigFloat>) {
+    out[(0, 1)] -= coeff * v[2];
+    out[(0, 2)] += coeff * v[1];
+    out[(1, 0)] += coeff * v[2];
+    out[(1, 2)] -= coeff * v[0];
+    out[(2, 0)] -= coeff * v[1];
+    out[(2, 1)] += coeff * v[0];
+}
+
+/// Add coeff * v.dot(v.T) to the output matrix.
+fn add_outer_prod(coeff: BigFloat, v: &Matrix3x1<BigFloat>, out: &mut Matrix3<BigFloat>) {
+    out[(0, 0)] += coeff * v[0] * v[0];
+    out[(0, 1)] += coeff * v[0] * v[1];
+    out[(0, 2)] += coeff * v[0] * v[2];
+    out[(1, 0)] += coeff * v[1] * v[0];
+    out[(1, 1)] += coeff * v[1] * v[1];
+    out[(1, 2)] += coeff * v[1] * v[2];
+    out[(2, 0)] += coeff * v[2] * v[0];
+    out[(2, 1)] += coeff * v[2] * v[1];
+    out[(2, 2)] += coeff * v[2] * v[2];
+}
+
+/// Compute the group commutator UVU^{-1}V^{-1}. This code leverages the fact that
+/// the matrices as SO(3) and U^{-1} = U^T.
+fn group_commutator(left: &Matrix3<BigFloat>, right: &Matrix3<BigFloat>) -> Matrix3<BigFloat> {
+    left * right * left.transpose() * right.transpose()
+}
+
+/// Compute the rotation axis of a SO(3) matrix.
+fn rotation_axis_from_so3(matrix: &Matrix3<BigFloat>, tol: BigFloat) -> Matrix3x1<BigFloat> {
+    let trace = matrix.trace();
+
+    // Try to obtain the matrix via Rodrigues formula. If the matrix is diagonal (or the Rodrigues
+    // approach fails, likely because the matrix was diagonal up to numerical error), determine
+    // the axis from the diagonal.
```

Comment:
I didn't understand this function (yet), can you provide the reference or explain the following computations? As we discussed, the Python version of this has problems, so I would really like to understand how this solves them.

Reply:
OOD, we discussed this offline